lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/Versions.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.base;
+
+import io.trino.spi.connector.ConnectorContext;
+import io.trino.spi.connector.ConnectorFactory;
+
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+
+public final class Versions
+{
+    private Versions() {}
+
+    /**
+     * Check if the SPI version of the Trino server matches exactly the SPI version the connector plugin was built for.
+     * Using plugins built for a different version of Trino may fail at runtime, especially if plugin author
+     * chooses not to maintain compatibility with older SPI versions, as happens for plugins maintained together with
+     * the Trino project.
+     */
+    public static void checkSpiVersion(ConnectorContext context, ConnectorFactory connectorFactory)
+    {
+        Optional<String> spiVersion = context.getSpiVersion();
+        String pluginVersion = connectorFactory.getClass().getPackage().getImplementationVersion();
+        if (pluginVersion == null) {
+            // Assume we're in tests. In tests, plugin version is unknown and SPI version may be known, e.g.
+            // when running single module's tests from maven.
+            return;
+        }
+
+        checkState(
```

Comment:
For example if one changed the engine in the very minor way, that does not affect the connector and now old plugins does not work while they could. I mean custom builds.

On the other hand this could be controlled on the plugin itself. So I am not sure about this.

Reply:
Plugging in a config toggle is complicated, because at the level the check is done, there is no config processing yet.

> For example if one changed the engine in the very minor way, that does not affect the connector and now old plugins does not work while they could.

"engine minor change" is subjective. This is not a scenario i would like to support. 

> I mean custom builds.

if someone has engine medications (a project fork), they would need to build plugins too.
However, if you manage a fork, building plugins is the most natural thing to do anyway

Reply:
@kokosing does this break the flow you wanted to achieve with single version backward SPI compatibility?
If we want to have `single version backward SPI compatibility`, then I think we need to account for that in the check below (connector supports current SPI or a single, incremented version)